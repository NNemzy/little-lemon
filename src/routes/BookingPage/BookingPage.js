import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import BookingPage from ".";

describe("BookingPage component", () => {
  test("renders successfully", () => {
    const { container } = render(
      <MemoryRouter>
        <BookingPage />
      </MemoryRouter>
    );
    console.log(container);
    const heading = screen.getByText("Booking Form");
    expect(heading).toBeInTheDocument();
    screen.debug();
  });
});

// test("loads without crashing");
// test("date input is in the document");
// test("date input has default value");
// test("date input value is current day");
// test("date input change loads without crashing");
// test("date input ");
