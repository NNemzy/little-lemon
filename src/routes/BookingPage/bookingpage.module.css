.container {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: w;
  height: 700px;
}
.container img {
  width: 700px;
}
.heading {
  margin-bottom: 100px;
}
.form {
  display: flex;
  align-items: center;
  flex-direction: column;
  width: 900px;
}
.form h1 {
  font-size: 3rem;
}

@media (max-width: 1000px) {
  .container img {
    width: 100%;
    height: 200px;
  }
  .form {
    padding: 50px;
  }
  .container {
    flex-direction: column;
  }
}
