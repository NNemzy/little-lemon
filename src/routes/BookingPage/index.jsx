import React, { useReducer, useState, useEffect } from "react";
import BookingForm from "../../components/BookingForm";
import { fetchAPI, submitAPI } from "./../../util/bookingApi/api";
import { useNavigate } from "react-router-dom";

import { reducer, updateTimes } from "./../../util/reducers/bookingReducer";

function initializeTimes() {
  return fetchAPI(new Date());
}

function BookingPage() {
  const navigate = useNavigate();
  const currentDate = new Date()
    .toLocaleDateString()
    .split("/")
    .reverse()
    .join("-");

  const [disableForm, setDisableForm] = useState(false);
  const [availableTimes, dispatch] = useReducer(reducer, initializeTimes());

  const [formValues, setFormValues] = useState({
    date: currentDate,
    time: availableTimes,
    guests: "1",
    occasion: "Birthday",
  });

  function updateTimes(data) {
    dispatch({ type: "DATE_CHANGE", payload: data });
    return;
  }

  useEffect(() => {}, []);

  const handleInputChange = (name) => (event) => {
    let inputValue = event.target.value;
    if (name === "date") {
      updateTimes(inputValue);
    }
    setFormValues({
      ...formValues,
      [name]: inputValue,
    });
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    setDisableForm(!disableForm);
    console.log("confirmed booking");
    // submitAPI() && navigate("/booking-confirmed");
  };

  return (
    <>
      <h1>Booking Form</h1>
      <BookingForm
        availableTimes={availableTimes}
        formValues={formValues}
        disableForm={disableForm}
        handleInputChange={handleInputChange}
        handleSubmit={handleSubmit}
      />
    </>
  );
}

export default BookingPage;
