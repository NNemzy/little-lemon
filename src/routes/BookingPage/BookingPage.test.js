import * as React from "react";
import "@testing-library/jest-dom";
import { createMemoryRouter, RouterProvider } from "react-router-dom";

import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { fetchAPI } from "./../../util/bookingApi/api";
import userEvent from "@testing-library/user-event";

// Component
import BookingPage from ".";

describe("Booking Page", () => {
  it("renders", () => {
    renderPage(<BookingPage />, "/booking");

    const headings = screen.getAllByRole("heading");
    expect(headings[0]).toHaveTextContent("Booking Form");
  });

  it("return the current date as default value", () => {
    renderPage(<BookingPage />, "/booking");
    const currentDate = new Date()
      .toLocaleDateString()
      .split("/")
      .reverse()
      .join("-");

    const dateInput = screen.getByLabelText("Choose Date:");

    expect(dateInput.value).toEqual(currentDate);
  });

  it("returns different times on different days", () => {
    renderPage(<BookingPage />, "/booking");
    const currentDateTimes = fetchAPI(new Date());
    const defaultDate = screen.getByLabelText("Choose Date:");
    const times = screen.getByLabelText("Choose Time:").children;

    const timeCopy = [...times];

    fireEvent.change(defaultDate, { target: { value: "2023-01-23" } });

    expect(currentDateTimes).not.toEqual(times);
    expect(times).not.toEqual(timesCopy);
  });

  it("");
});

function renderPage(Component, path) {
  const routes = [{ path, element: Component }];
  const initialEntries = [path];
  const initialIndex = 0;
  const router = getRouter(routes, initialEntries, initialIndex);
  return render(<RouterProvider router={router} />);
}

function getRouter(routes, initialEntries, initialIndex) {
  return createMemoryRouter(routes, {
    initialEntries,
    initialIndex,
  });
}
