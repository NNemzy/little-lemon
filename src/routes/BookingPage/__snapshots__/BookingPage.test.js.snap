// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking Page returns different times on different days 1`] = `
<div>
  <h1>
    Booking Form
  </h1>
  <form
    class="form"
  >
    <div>
      <label
        for="res-date"
      >
        Choose Date:
      </label>
      <input
        id="res-date"
        max="2023-02-06"
        min="2023-01-06"
        required=""
        type="date"
        value="2023-01-23"
      />
    </div>
    <div>
      <label
        for="res-time"
      >
        Choose Time:
      </label>
      <select
        id="res-time"
      >
        <option>
          17:00
        </option>
        <option>
          17:30
        </option>
        <option>
          19:00
        </option>
        <option>
          19:30
        </option>
        <option>
          21:30
        </option>
        <option>
          22:00
        </option>
        <option>
          23:30
        </option>
      </select>
    </div>
    <div>
      <label
        for="guest"
      >
        Number of Guests: 
      </label>
      <input
        id="guests"
        max="10"
        min="1"
        placeholder="1"
        type="number"
        value="1"
      />
    </div>
    <label
      for="occasion"
    >
      Occasion: 
    </label>
    <select
      id="occasion"
    >
      <option>
        Birthday
      </option>
      <option>
        Anniversary
      </option>
    </select>
    <div>
      <input
        type="submit"
        value="Make your reservation"
      />
    </div>
  </form>
</div>
`;
